pkg/schemas/service.go

```diff
@@ -0,0 +1,264 @@
+}
+
+// TODO @shipperizer verify during integration test if this is actually the format
+type KratosError struct {
+	Error *kClient.GenericError `json:"error,omitempty"`
+}
+
+type Service struct {
+	cmName      string
+	cmNamespace string
+
+	k8s    coreV1.CoreV1Interface
+	kratos kClient.IdentityApi
+
+	tracer  trace.Tracer
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (s *Service) parseError(ctx context.Context, r *http.Response) *kClient.GenericError {
+	ctx, span := s.tracer.Start(ctx, "schemas.Service.parseError")
+	defer span.End()
+
+	gerr := KratosError{Error: kClient.NewGenericErrorWithDefaults()}
+
+	defer r.Body.Close()
+	body, _ := io.ReadAll(r.Body)
+
+	if err := json.Unmarshal(body, &gerr); err != nil {
+		gerr.Error.SetMessage("unable to parse kratos error response")
+		gerr.Error.SetCode(http.StatusInternalServerError)
+	}
+
+	return gerr.Error
+}
+
+func (s *Service) ListSchemas(ctx context.Context, page, size int64) (*IdentitySchemaData, error) {
+	ctx, span := s.tracer.Start(ctx, "schemas.Service.ListSchemas")
+	defer span.End()
+
+	schemas, rr, err := s.kratos.ListIdentitySchemasExecute(
+		s.kratos.ListIdentitySchemas(ctx).Page(page).PerPage(size),
+	)
+
+	data := new(IdentitySchemaData)
+
+	if err != nil {
+		s.logger.Error(err)
+		data.Error = s.parseError(ctx, rr)
+	}
+
+	data.IdentitySchemas = schemas
+
+	// TODO @shipperizer check if schemas is defaulting to empty slice inside kratos-client
+	if data.IdentitySchemas == nil {
+		data.IdentitySchemas = make([]kClient.IdentitySchemaContainer, 0)
+	}
+
+	return data, err
+}
+
+func (s *Service) GetSchema(ctx context.Context, ID string) (*IdentitySchemaData, error) {
+	ctx, span := s.tracer.Start(ctx, "schemas.Service.GetSchema")
+	defer span.End()
+
+	schema, rr, err := s.kratos.GetIdentitySchemaExecute(
+		s.kratos.GetIdentitySchema(ctx, ID),
+	)
+
+	data := new(IdentitySchemaData)
+
+	if err != nil {
+		s.logger.Error(err)
+		data.Error = s.parseError(ctx, rr)
+	}
+
+	if schema != nil {
+		data.IdentitySchemas = []kClient.IdentitySchemaContainer{
+			{Schema: schema, Id: &ID},
+		}
+	} else {
+		data.IdentitySchemas = []kClient.IdentitySchemaContainer{}
+	}
+
+	return data, err
+}
+
+func (s *Service) EditSchema(ctx context.Context, ID string, data *kClient.IdentitySchemaContainer) (*IdentitySchemaData, error) {
+	ctx, span := s.tracer.Start(ctx, "schemas.Service.EditSchema")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	i := new(IdentitySchemaData)
+
+	schemas := s.schemas(cm.Data)
+
+	if _, ok := schemas[ID]; !ok {
+		i.IdentitySchemas = []kClient.IdentitySchemaContainer{}
+
+		return i, fmt.Errorf("schema with ID %s not found", ID)
+	}
+
+	rawSchema, err := json.Marshal(data.Schema)
+
+	if err != nil {
+		return nil, err
+	}
+
+	cm.Data[ID] = string(rawSchema)
```

Comment:
https://github.com/canonical/identity-platform-admin-ui/pull/43/files/16ac14fae1ea4db3c6379a1704478ad881b942f3#diff-1dcb448ddab2bbe36e0d7103bb601799f3026828c8e83940788dba1c67712ab4R130 this would prevent it

Reply:
although good catch as here https://github.com/canonical/identity-platform-admin-ui/pull/43/files/16ac14fae1ea4db3c6379a1704478ad881b942f3#diff-1dcb448ddab2bbe36e0d7103bb601799f3026828c8e83940788dba1c67712ab4R130 it will blow up kratos